### ISSUE TITLE
Replace batcher sleeps with signals

### ISSUE THREAD

**Issue by GGP1** [wazuh/wazuh#27405]
## Description

During https://github.com/wazuh/wazuh/issues/27172, we detected that the batcher performs some sleep operations when receiving items from the multiplexer queue

https://github.com/wazuh/wazuh/blob/8d532a33c5907ea6961fb194fe7a87a398c349b4/framework/wazuh/core/batcher/batcher.py#L57-L63

and when receiving responses from the demultiplexer queue

https://github.com/wazuh/wazuh/blob/8d532a33c5907ea6961fb194fe7a87a398c349b4/framework/wazuh/core/batcher/client.py#L58-L62

Using infinite loops with sleeps add unnecessary latency to the workflow and consumes resources when trying to obtain items.

We should replace them with an event-driven approach in which we block/unblock the execution based on event signals like we currently do in the Communications API internal commands manager

https://github.com/wazuh/wazuh/blob/8d532a33c5907ea6961fb194fe7a87a398c349b4/apis/comms_api/core/commands.py#L68-L77

## Checks

The following elements have been updated or reviewed (should also be checked if no modification is required):
- [ ] Tests (unit tests, API integration tests).
- [ ] Changelog.
- [ ] Documentation.
- [ ] Integration test mapping (using `api/test/integration/mapping/_test_mapping.py`).

**Comment by RamosFe** [comment]
## Update

When adding the signals to remove waits, I encountered the following error. I continue developing and seeing how to find a way around this problem.

```python
wazuh-manager  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/managers.py", line 833, in _callmethod
wazuh-manager  |     raise convert_to_error(kind, result)
wazuh-manager  | RuntimeError: Condition objects should only be shared between processes through inheritance
```

**Comment by fdalmaup** [comment]
On hold since @RamosFe is OOO.

**Comment by RamosFe** [comment]
## Update 06/1

Adding a new `multiprocessing.Manager` doesn't solve the problem and causes the following error due to the necessity of sharing the `multiprocessing.Manager` between multiple processes:

```console
wazuh-worker1  |     batcher_mux_demux_manager = MuxDemuxManager()
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/site-packages/wazuh/core/batcher/mux_demux.py", line 344, in __init__
wazuh-worker1  |     self.queue = self.manager.MuxDemuxQueue(
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/managers.py", line 723, in temp
wazuh-worker1  |     token, exp = self._create(typeid, *args, **kwds)
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/managers.py", line 608, in _create
wazuh-worker1  |     id, exposed = dispatch(conn, None, 'create', (typeid,)+args, kwds)
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/managers.py", line 89, in dispatch
wazuh-worker1  |     c.send((id, methodname, args, kwds))
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/connection.py", line 206, in send
wazuh-worker1  |     self._send_bytes(_ForkingPickler.dumps(obj))
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
wazuh-worker1  |     cls(buf, protocol).dump(obj)
wazuh-worker1  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/process.py", line 353, in __reduce__
wazuh-worker1  |     raise TypeError(
wazuh-worker1  | TypeError: Pickling an AuthenticationString object is disallowed for security reasons
```

I replicated the same code that we have in the `CommandsManager`, as shown below:

```python

    async def receive_from_demux(self, packet_id: int) -> Packet:
        """Retrieve and remove a response from the dictionary for a given unique identifier.

        Parameters
        ----------
        packet_id : int
            Unique identifier of the response.

        Returns
        -------
        dict
            Indexer response.
        """
        if packet_id not in self.responses:
            event = asyncio.Event()
            self._subscriptions.update({packet_id: event})

            signaled = await event.wait()
            self._subscriptions.pop(packet_id, None)
            if not signaled:
                return Packet()

        response = self.responses[packet_id]
        del self.responses[packet_id]
        return response

    def internal_store_response(self, packet: Packet):
        """Update the responses dictionary with the item content.

        Parameters
        ----------
        packet : Packet
            Packet whose content will be added to the response dictionary.
        """
        if packet.id not in self.responses:
            self.responses[packet.id] = packet
        else:
            # Using self.responses[packet.id].append() doesn't work because
            # it doesn't hold the reference of nested objects
            response = self.responses[packet.id]
            response.items.extend(packet.items)
            self.responses[packet.id] = response

        if packet.id in self._subscriptions:
            self._subscriptions[packet.id].set()

```

But the following error occurs:

```console
wazuh-manager  | Unserializable message: Traceback (most recent call last):
wazuh-manager  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/managers.py", line 308, in serve_client
wazuh-manager  |     send(msg)
wazuh-manager  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/connection.py", line 206, in send
wazuh-manager  |     self._send_bytes(_ForkingPickler.dumps(obj))
wazuh-manager  |   File "/usr/share/wazuh-server/framework/python/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
wazuh-manager  |     cls(buf, protocol).dump(obj)
wazuh-manager  | TypeError: cannot pickle 'coroutine' object
wazuh-manager  | 
```

I will continue investigating why the error appears in the `MuxDemuxQueue` but not in the `CommandsManager`.

**Comment by RamosFe** [comment]
## Update 8/1

The error mentioned in https://github.com/wazuh/wazuh/issues/27405#issuecomment-2573866569 was caused by using `async` in the `receive_from_demux` function. The `async` function returns a `coroutine` that must be passed between the two processes, which led to the error.

To resolve this, I replaced the `asyncio` events with `multiprocessing` events, setting the timeout to match the endpoint timeout. Using these types of events allows requests to block the main event loop of the API. To prevent this blockage, I utilized the `asyncio.to_thread` function, which helps run blocking operations in a separate thread without halting the main event loop. As explained in the [documentation](https://docs.python.org/3/library/asyncio-task.html):

> Directly calling blocking_io() in any coroutine would block the event loop for its duration. Instead, by using asyncio.to_thread(), we can execute it in a separate thread without blocking the event loop.

### UT

The UTs related to the Batcher were updated:

```console
(env) federamos@pop-os:~/Documents/Wazuh/Repositories/wazuh$ sudo -E PYTHONPATH=framework:apis:api env/bin/python3 -m pytest framework/wazuh/core/batcher/
========================================================================================================== test session starts ==========================================================================================================
platform linux -- Python 3.10.12, pytest-7.3.1, pluggy-1.5.0
rootdir: /home/federamos/Documents/Wazuh/Repositories/wazuh/framework
configfile: pytest.ini
plugins: asyncio-0.18.1, html-2.1.1, metadata-3.1.1, cov-4.1.0, tavern-1.23.5, anyio-4.1.0, trio-0.8.0
asyncio: mode=auto
collected 34 items                                                                                                                                                                                                                      

framework/wazuh/core/batcher/tests/test_buffer.py .............                                                                                                                                                                   [ 38%]
framework/wazuh/core/batcher/tests/test_client.py ..                                                                                                                                                                              [ 44%]
framework/wazuh/core/batcher/tests/test_mux_demux.py ................                                                                                                                                                             [ 91%]
framework/wazuh/core/batcher/tests/test_timer.py ...                                                                                                                                                                              [100%]

========================================================================================================== 34 passed in 0.36s ===========================================================================================================
```

### Execution

```console
wazuh-manager  | 2025/01/08 18:25:45 DEBUG: [Communications API] Request headers: {'authorization': 'Bearer eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ3YXp1aCIsImF1ZCI6IldhenVoIENvbW11bmljYXRpb25zIEFQSSIsImlhdCI6MTczNjM2MDczMywiZXhwIjoxNzM2MzYxNjMzLCJ1dWlkIjoiMDE5MTJkMDctYjliNC03NTI4LWJkYWQtMTVkYTkwMmQ2NTFjIn0.c6uURroEqqwBk7iFpZyGATflddgE0g0bVlYU7NTWu9mut-AK3nf0QmHn8dxmzUHftNk8MjtDSYfBAPkAC-Mgeg', 'content-type': 'text/plain', 'user-agent': 'PostmanRuntime/7.32.3', 'accept': '*/*', 'cache-control': 'no-cache', 'postman-token': '89425564-34a4-4ae1-9df3-2611e2ea45e1', 'host': 'localhost:27000', 'accept-encoding': 'gzip, deflate, br', 'connection': 'keep-alive', 'content-length': '699'}
wazuh-manager  | 2025/01/08 18:25:45 DEBUG: [Communications API] Request body stream: {"agent": {"id": "0a96a0ab-5bef-415c-bb3c-ea3e294215a0", "name": "example", "type": "endpoint", "version": "5.0.0", "groups": ["group1"], "host": {"hostname": "myhost", "os": {"name": "Amazon Linux 2", "platform": "Linux"}, "ip": ["192.168.1.2"], "architecture": "x86_64"}}}
wazuh-manager  | {"module": "fim", "operation": "create", "id": "1"}
wazuh-manager  | {"file": {"attributes": ["string"], "name": "string", "path": "string", "gid": 0, "group": "string", "inode": "string", "mtime": "2024-10-30T14:59:31.558Z", "mode": "string", "size": 0, "target_path": "string", "type": "string", "uid": 0, "owner": "string", "hash": {"md5": "string", "sha1": "string", "sha256": "string"}}, "registry": {"key": "string","value": "string"}}
wazuh-manager  | 2025/01/08 18:25:45 INFO: [Communications API] (01912d07-b9b4-7528-bdad-15da902d651c) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.216s: 200
```

**Comment by RamosFe** [comment]
## Update

The following results were obtained after performing a test using an environment with 20 agents.

| Issue | Median | Min | Max |  Sum | Difference |
| -- | -- | -- | -- | -- | -- |
| With sleep | 768 | 160 | 820 | 13501 |  |
| With signals | 674 | 77 | 769 | 11615 | **-16.24%** |

<details>
<summary>Batcher config</summary>

```console
communications_api:
  host: '0.0.0.0'
  batcher:
    max_elements: 9999999999999
    max_size: 99999999999999
    wait_time: 0.1
```

</details>


<details>
<summary>Logs</summary>

```console
2025/01/15 12:15:23 INFO: [Communications API] Stateful events batch. Packets: 1. Items: 128
2025/01/15 12:15:23 INFO: [Communications API] (9977ca5a-36e6-41f3-9f2f-91d0af13b1c6) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.077s: 200
2025/01/15 12:15:23 INFO: [Communications API] Stateful events batch. Packets: 1. Items: 128
2025/01/15 12:15:23 INFO: [Communications API] (987cee73-611b-4589-8adb-8930fc3fa495) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.136s: 200
2025/01/15 12:15:24 INFO: [Communications API] Stateful events batch. Packets: 13. Items: 1664
2025/01/15 12:15:24 INFO: [Communications API] (c9c4e1b9-4666-47a7-996b-ed47d952c66e) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.406s: 200
2025/01/15 12:15:25 INFO: [Communications API] (003a67fb-3e19-4bb0-9fc6-4101534ed820) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.346s: 200
2025/01/15 12:15:24 INFO: [Communications API] Stateful events batch. Packets: 5. Items: 640
2025/01/15 12:15:25 INFO: [Communications API] (e32bce5c-5fee-4d27-a443-48ea55660e83) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.379s: 200
2025/01/15 12:15:25 INFO: [Communications API] (edfa4a9d-cab3-4775-9077-0d18347bda83) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.716s: 200
2025/01/15 12:15:26 INFO: [Communications API] (0fab5acf-b3d3-4e22-be66-3c7af9ba65dd) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.671s: 200
2025/01/15 12:15:26 INFO: [Communications API] (fa776051-e256-4de6-8268-33ddfaba0d4f) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.677s: 200
2025/01/15 12:15:26 INFO: [Communications API] (4f0624c2-ebea-40dc-8b5c-5cfc43303f66) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.661s: 200
2025/01/15 12:15:27 INFO: [Communications API] (150d0ef4-c530-451e-a0ea-56be2ac8a99e) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.691s: 200
2025/01/15 12:15:27 INFO: [Communications API] (607a089b-204f-48a4-8161-b1e77f60e68c) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.619s: 200
2025/01/15 12:15:27 INFO: [Communications API] (5f93fd3a-e605-4c6e-9641-b65e238d69da) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.602s: 200
2025/01/15 12:15:28 INFO: [Communications API] (175bc102-dc24-41c4-88d1-43847e58b13f) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.692s: 200
2025/01/15 12:15:28 INFO: [Communications API] (6f53946c-6315-42e9-a581-f03a9b988ea9) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.769s: 200
2025/01/15 12:15:28 INFO: [Communications API] (6a09a421-7cfe-4d64-af29-656ebe8fe421) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.716s: 200
2025/01/15 12:15:29 INFO: [Communications API] (51f42d8b-e892-4892-9362-5579f0171083) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.653s: 200
2025/01/15 12:15:29 INFO: [Communications API] (33d3e54a-dfd0-4dbd-9a39-4cd209a037fe) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.715s: 200
2025/01/15 12:15:29 INFO: [Communications API] (f72d7fc6-d470-4fa0-b748-4e95098a6e76) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.693s: 200
2025/01/15 12:15:29 INFO: [Communications API] (8800324c-d86f-4556-a674-a11eeeb44a6f) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.705s: 200
2025/01/15 12:15:30 INFO: [Communications API] (d1352479-5c4b-4d81-82a5-01e07be93482) "POST /api/v1/events/stateful" with parameters {} and body {} done in 0.691s: 200
```

</details>
